### PR TITLE
joint_state_broadcaster to use realtime tools

### PR DESCRIPTION
**Pull request by bmagyar**

### CODE REVIEW

File: joint_state_broadcaster/package.xml [hunk 1]
```diff
@@ -6,7 +6,6 @@
   <maintainer email="bence.magyar.robotics@gmail.com">Bence Magyar</maintainer>
   <maintainer email="denis@stogl.de">Denis Stogl</maintainer>
   <maintainer email="jordan.palacios@pal-robotics.com">Jordan Palacios</maintainer>
-  <maintainer email="karsten@openrobotics.org">Karsten Knese</maintainer>
```
Comment (destogl) [inline joint_state_broadcaster/package.xml:9]: Let's change list of maintainers in a separate PR and in all packages at once.

File: joint_state_broadcaster/src/joint_state_broadcaster.cpp [hunk 2]
```diff
@@ -298,37 +308,40 @@ controller_interface::return_type JointStateBroadcaster::update(
...
-      dynamic_joint_state_msg_.interface_values[joint_index].values[interface_index] =
+        dynamic_joint_state_msg.interface_values[joint_index].interface_names[interface_index];
+      dynamic_joint_state_msg.interface_values[joint_index].values[interface_index] =
         name_if_value_mapping_[name][interface_name];
     }
   }
   // publish
-  joint_state_publisher_->publish(joint_state_msg_);
-  dynamic_joint_state_publisher_->publish(dynamic_joint_state_msg_);
+  realtime_joint_state_publisher_->unlockAndPublish();
```
Comment (destogl) [inline joint_state_broadcaster/src/joint_state_broadcaster.cpp:343]: You forgot to lock the variables.

File: joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp [hunk 3]
```diff
@@ -94,14 +95,16 @@ class JointStateBroadcaster : public controller_interface::ControllerInterface
   //  we store the name of joints with compatible interfaces
   std::vector<std::string> joint_names_;
   std::shared_ptr<rclcpp::Publisher<sensor_msgs::msg::JointState>> joint_state_publisher_;
-  sensor_msgs::msg::JointState joint_state_msg_;
+  std::shared_ptr<realtime_tools::RealtimePublisher<sensor_msgs::msg::JointState>>
+    realtime_joint_state_publisher_ = nullptr;
```
Comment (destogl) [inline joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:99]: Not sure about 'realtime' prefix though...

I would call the normal one with 'pub_' suffix and this one 'publisher_'


One way or another we should sync naming in all broadcasters we have.

**Review by destogl — CHANGES_REQUESTED** [review]

File: joint_state_broadcaster/package.xml (hunk 1 above)
Comment (bmagyar) [reply joint_state_broadcaster/package.xml:9]: Yessir

File: joint_state_broadcaster/src/joint_state_broadcaster.cpp (hunk 2 above)
Comment (bmagyar) [reply joint_state_broadcaster/src/joint_state_broadcaster.cpp:343]: Oh no! You are right

File: joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp (hunk 3 above)
Comment (bmagyar) [reply joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:99]: I didn't have a better idea... I actually want to discuss the realtime publisher API, I'm r ally.not happy with it

File: joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp [hunk 4]
```diff
@@ -100,14 +101,16 @@ class JointStateBroadcaster : public controller_interface::ControllerInterface
...
```
Comment (destogl) [inline joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:105]: It is really needed to set the initial value to `nullptr`?

If so, we should add this to all controllers (`good-first-issue`). If needed please open follow-up issue.

File: joint_state_broadcaster/src/joint_state_broadcaster.cpp [hunk 5]
```diff
@@ -342,37 +352,45 @@ controller_interface::return_type JointStateBroadcaster::update(
       state_interface.get_value());
   }
 
-  joint_state_msg_.header.stamp = time;
-  dynamic_joint_state_msg_.header.stamp = time;
-
-  // update joint state message and dynamic joint state message
-  for (size_t i = 0; i < joint_names_.size(); ++i)
+  if (realtime_joint_state_publisher_ && realtime_dynamic_joint_state_publisher_->trylock())
```
Comment (destogl) [inline joint_state_broadcaster/src/joint_state_broadcaster.cpp:355]: It looks like you mixed the publishers?

```suggestion
  if (realtime_joint_state_publisher_ && realtime_joint_state_publisher_->trylock())
```

**Review by destogl — APPROVED** [review]
It looks good, but please check the comment about mixed publishers!

Thanks!

File: joint_state_broadcaster/src/joint_state_broadcaster.cpp (hunk 5 above)
Comment (bmagyar) [reply joint_state_broadcaster/src/joint_state_broadcaster.cpp:355]: nice catch, thanks! :D

File: joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp (hunk 4 above)
Comment (bmagyar) [reply joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:105]:
```suggestion
    realtime_joint_state_publisher_;
```
du has recht, not strictly needed

File: joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp [hunk 6]
```diff
@@ -100,14 +101,16 @@ class JointStateBroadcaster : public controller_interface::ControllerInterface
...
 
   //  For the DynamicJointState format, we use a map to buffer values in for easier lookup
   //  This allows to preserve whatever order or names/interfaces were initialized.
   std::unordered_map<std::string, std::unordered_map<std::string, double>> name_if_value_mapping_;
   std::shared_ptr<rclcpp::Publisher<control_msgs::msg::DynamicJointState>>
     dynamic_joint_state_publisher_;
-  control_msgs::msg::DynamicJointState dynamic_joint_state_msg_;
+  std::shared_ptr<realtime_tools::RealtimePublisher<control_msgs::msg::DynamicJointState>>
+    realtime_dynamic_joint_state_publisher_ = nullptr;
```
Comment (bmagyar) [inline joint_state_broadcaster/include/joint_state_broadcaster/joint_state_broadcaster.hpp:113]:
```suggestion
    realtime_dynamic_joint_state_publisher_;
```

**Comment by destogl** [comment]
It seems you have to also adjust tests too because you deleted an internal variable for storing messages.

**Review by destogl — APPROVED** [review]
Looking good. Let's just wait for the CI.

**Pull request merged**